### ISSUE TITLE
create connection between host AP machine and qemu image as STA

### ISSUE THREAD

**Issue by Dvergatal** [intrig-unicamp/mininet-wifi#480]
As described in subject I am willing to create a test environment in our project for testing wifi implementation. Currently we are using qemu which is being loaded in docker. The docker is already configured for building mininet-wifi but question is how can I configure docker host machine to behave as AP and the qemu image as STA and made the AP visible by the STA?

**Comment by ramonfontes** [comment]
> how can I configure docker host machine to behave as AP

https://github.com/intrig-unicamp/mininet-wifi/blob/master/examples/sta_ap_mode.py
https://github.com/ramonfontes/containernet

> qemu image as STA

The only thing you can do (I guess) is to create a virtual interface for each VM - but I think this is not possible. Honestly, I see no reason for creating virtual interface interfaces for VMs if you can do with docker (e.g. by using containernet)

**Comment by Dvergatal** [comment]
Hi @ramonfontes I'm sorry for long response but I was handling other works. 

> The only thing you can do (I guess) is to create a virtual interface for each VM - but I think this is not possible. Honestly, I see no reason for creating virtual interface interfaces for VMs if you can do with docker (e.g. by using containernet)

Because I want to test if my code written for establishing connection from my image as STA to AP is working without issues.

Additionally is it possible to configure containernet to have one STA behaving as client and second one as AP and they will be both visible for each other?

**Comment by ramonfontes** [comment]
> Additionally is it possible to configure containernet to have one STA behaving as client and second one as AP and they will be both visible for each other?

Are you talking about https://github.com/intrig-unicamp/mininet-wifi/blob/master/examples/sta_ap_mode.py#L37?

**Comment by Dvergatal** [comment]
@ramonfontes Ok so this is in just one docker container, which have two wireless interfaces?

**Comment by ramonfontes** [comment]
I didn't get you point.. You may have that setup with containers. Each node will be running in a particular container. If you one node must have multiple wireless interfaces this is up to you.

**Comment by Dvergatal** [comment]
What I want to achieve is to establish connection from one interface to another with the usage of code which is available on the qemu image and for that I need to share one virtual wifi device with qemu and second one needs to be an access point which will be visible fort the client one. Do you understand know?

**Comment by Dvergatal** [comment]
OK, but what do you mean by `veths`? virtual ethernet? And does this solution will allow me to have possibility of having two wireless interfaces visible by each other - meaning e.g. that creating on first one an access point and on the second one scanning for available access points, will result with an output with this access point being on the list.

**Comment by Dvergatal** [comment]
@ramonfontes on the uml diagram below this is exactly what I would like to achieve
![docker-qemu](https://user-images.githubusercontent.com/72452190/215736056-ccfeeaf6-f420-4edc-9f0f-ef23f7879849.png)
Both wlan interfaces would be available from minitnet-wifi. The wlan0 would be configured as AP and wlan1 would be an STA available from QEMU image. Now form QEMU image with this wlan1 interface I would be able to scan area and see that there is an access point with ssid on wlan0 and establish a connection to it.

**Comment by ramonfontes** [comment]
Running station with inNamespace = False and attach the wifi interface to the QEMU image should support your goals.

**Comment by Dvergatal** [comment]
@ramonfontes OK i will give a shot and come back to you with an update

**Comment by Dvergatal** [comment]
OK I'm back @ramonfontes I have configured the `docker-compose.yml` and `Dockerfile` for mininet-wifi and installed it with below commands:
```bash
git clone https://github.com/intrig-unicamp/mininet-wifi.git
util/install.sh -Wlnfv
```
and taken your example code from [sta_ap_mode.py](https://github.com/intrig-unicamp/mininet-wifi/blob/master/examples/sta_ap_mode.py) because I will need to create these interfaces but unfortunately I'm getting an error that:
```bash
root@latitude:/work# python sta_ap_mode.py
Traceback (most recent call last):
  File "/work/sta_ap_mode.py", line 7, in <module>
    from mininet.log import setLogLevel, info
ModuleNotFoundError: No module named 'mininet'
```
Mininet-wifi package says that it is `Installed /usr/local/lib/python3.10/dist-packages/mininet_wifi-2.6-py3.10.egg` so it should work...

**Comment by ramonfontes** [comment]
I think something did not go as expected during the mn-wifi installation. The mininet module exists.

**Comment by Dvergatal** [comment]
> I think something did not go as expected during the mn-wifi installation. The mininet module exists.

Yeah that is what I was thinking but I can give you logs from building process need few minutes for that BRB

**Comment by Dvergatal** [comment]
I'm getting 'Failed to disable unit, unit openvswitch-testcontroller.service does not exist.` and that is all at the end of the log file...

**Comment by Dvergatal** [comment]
@ramonfontes but mn command is working

**Comment by ramonfontes** [comment]
I think there is nothing I can do. This is a particular issue with your system.

**Comment by Dvergatal** [comment]
@ramonfontes I think I know what's the issue:
```python3
>>> import sys
>>> print(sys.path)
['', '/usr/lib/python310.zip', '/usr/lib/python3.10', '/usr/lib/python3.10/lib-dynload', '/opt/venv/lib/python3.10/site-packages']
```
and the install path is set to /usr/local/lib is it possible to re-define it to the script?

**Comment by Dvergatal** [comment]
OK now it is working :)

**Comment by Dvergatal** [comment]
@ramonfontes ok now my question is if I'm having the above example [sta_ap_mode.py](https://github.com/intrig-unicamp/mininet-wifi/blob/master/examples/sta_ap_mode.py) I would like to make a test that i will see these virtual interfaces running command e.g. ifconfig is that possible?

**Comment by Dvergatal** [comment]
@ramonfontes so my finall question that in order to create my configuration I need to re-write this python code to create AP and STA wifi interfaces and

> Running station with inNamespace = False and attach the wifi interface to the QEMU image should support your goals.

right?

**Comment by ramonfontes** [comment]
> OK now it is working :)

Could you please tell us the solution?

**Comment by Dvergatal** [comment]
@ramonfontes I've made an export for `PYTHONPATH` variable in shell with path `/usr/local/lib/python3.10/dist-packages`

**Comment by Dvergatal** [comment]
@ramonfontes yeah it is working finally and `nmcli device wifi list` command shows me the proper output:
```bash
IN-USE  BSSID              SSID      MODE   CHAN  RATE        SIGNAL  BARS  SECURITY 
        02:00:00:00:01:00  ap1-ssid  Infra  1     130 Mbit/s  44      **    --       

IN-USE  BSSID              SSID      MODE   CHAN  RATE        SIGNAL  BARS  SECURITY 
        02:00:00:00:02:00  ap2-ssid  Infra  6     130 Mbit/s  44      **    --       
```

**Comment by Dvergatal** [comment]
@ramonfontes btw. shouldn't be both networks visible by each other interfaces?

**Comment by ramonfontes** [comment]
No sure what do you mean with visible by each other interfaces. It would be great if elaborate the question a bit more.

**Comment by Dvergatal** [comment]
> No sure what do you mean with visible by each other interfaces. It would be great if you elaborate the question a bit more.

Access point with ssid `ap1-ssid` should see on interface name `sta1-wlan0` an  access point with ssid `ap2-ssid ` and vice versa

**Comment by ramonfontes** [comment]
Two stations with wireless interfaces working on master mode see each other? Why?

**Comment by Dvergatal** [comment]
> Two stations with wireless interfaces working on master mode see each other? Why?

What does it mean that they are on master mode?

**Comment by ramonfontes** [comment]
> What does it mean that they are on master mode?

Well, this is a 802.11 matter. I suggest you to take a read about it.

**Comment by Dvergatal** [comment]
I always thought that if I'm setting Access Point on an interface I'm still able to do the scan and see others and basically I'm doing it with `nmcli device wifi list ifname sta1-wlan0` but what is confusing me is that on ` sta1-wlan0` interface on which is set an Access Point with ssid `ap1-ssid` I see only `ap1-ssid` and I hoped to see also `ap2-ssid`

I do not know how does mininet-wifi works and that is why I'm asking, besides as I said I wanted to achieve the approach given previously on the uml diagram and to be honest I'm doing it using your examples

**Comment by Dvergatal** [comment]
@ramonfontes maybe I'll try my best and explain as easy as I can with a common example. If I'm having 2 wifi routers which both are having APs configured on their wifi interfaces I am able to make a scan on these interfaces and see available networks for both of them and here with this I've got the feeling that these `sta_ap_mode.py` example I've got the feeling that both interfaces are somehow separated....

**Comment by ramonfontes** [comment]
Mininet-wifi doesn't recreate the mac80211 stack. If you cannot perform scanning this may be related to IEEE 80211 itself, hostapd constraints and even wireless tools such as iw/iwconfig. 

I cannot perform a scanning for you right now but I invite you to either take a look at their documentation or issues on the Internet. 

Ps. I don't know why you using nmcli and not iw!

**Comment by Dvergatal** [comment]
@ramonfontes the same is with iw and I can perform scanning but both interfaces, APs are not being seen by each other

**Comment by ramonfontes** [comment]
Please help me and others. Tell us the command you use with the output.

**Comment by Dvergatal** [comment]
Ok I'm running the code from `sta_ap_mode.py` with commented line https://github.com/intrig-unicamp/mininet-wifi/blob/master/examples/sta_ap_mode.py#L71 and than i see with `ifconfig` command that I have 2 interfaces `sta1-wlan0` and `sta2-wlan0` and if I run `iw sta1-wlan0 scan` I can only see `ap1-ssid` which is set on `sta1-wlan0` but I think that I should also see `ap2-ssid` which is set on `sta2-wlan0`

**Comment by Dvergatal** [comment]
Or maybe I am reading the code completely wrong, let me put it this way if I'm calling `ap1 = net.addStation('ap1',  mac='02:00:00:00:01:00', ip='192.168.0.10/24', position='40,60,0')` is it being set on `sta1 = net.addStation('sta1', mac='00:00:00:00:00:01', ip='192.168.0.1/24', position='20,60,0')` or is it set on completly different interface?

**Comment by ramonfontes** [comment]
Isn't ap2 too far away to be reached or detected by sta1?

**Comment by Dvergatal** [comment]
@ramonfontes let's get one thing clear first. Is the setting of ap1 on sta1 or is it being created on completely different interface?

**Comment by ramonfontes** [comment]
Closing due to no new interactions.
